### ISSUE TITLE
Key type inference doesn't work if the data generic is provided

### ISSUE THREAD

**Issue by shuding** [vercel/swr#1518]
# Bug report

## Description / Observed Behavior

```ts
useSWR<Data>('/api', (endpoint) => sendRequestFor(endpoint))
                      ﹌﹌﹌﹌﹌ "Parameter 'endpoint' implicitly has an 'any' type.ts(7006)"
```

## Expected Behavior

It would be great if we can workaround this.

## Repro Steps / Code Example

N/A

## Additional Context

SWR 1.1.0-beta.3.

**Comment by shuding** [comment]
Update: we need partial inference for this, AFAIK it's still not supported by TS yet (https://github.com/microsoft/TypeScript/issues/26242).

**Comment by shuding** [comment]
Closed by #1640.
